### ISSUE TITLE
Problems with loading custom environment

### ISSUE THREAD

**Issue by BarisYazici** [atavakol/action-branching-agents#5]
I tried with generic environments such as "Humanoid-v2" and the save and load features work without an error. Now I am experimenting with action branching with a custom gym environment. My model is saved without a problem after training. However, I receive error in the loading process when I run the enjoy_continous.py.  I share the error message below

**System** 
- python = 3.6.10
- gym = 0.15.4  
- tensorflow-gpu = 1.12
- tensorflow = 1.15
- cloudpickle = 1.2.1  
- dill = 0.3.1.1  

_2020-05-14 14:57:12.020087: W tensorflow/core/framework/op_kernel.cc:1273] OP_REQUIRES failed at save_restore_v2_ops.cc:184 : Not found: Key Adam/beta_1 not found in checkpoint
Traceback (most recent call last):
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.NotFoundError: Key Adam/beta_1 not found in checkpoint
         [[{{node save/RestoreV2}} = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_INT64, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1546, in restore
    {self.saver_def.filename_tensor_name: save_path})
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.NotFoundError: Key Adam/beta_1 not found in checkpoint
         [[node save/RestoreV2 (defined at /home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/common/tf_util.py:301)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_INT64, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

Caused by op 'save/RestoreV2', defined at:
  File "action-branching-agents/agents/bdq/enjoy_continuous.py", line 84, in <module>
    main(args)
  File "action-branching-agents/agents/bdq/enjoy_continuous.py", line 53, in main
    act = deepq.load(args.model_dir)
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/deepq/procedure_continuous_tasks.py", line 77, in load
    return ActWrapper.load(path, num_cpu=num_cpu)
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/deepq/procedure_continuous_tasks.py", line 38, in load
    U.load_state(os.path.join(td, "model"))
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/common/tf_util.py", line 301, in load_state
    saver = tf.train.Saver()
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1102, in __init__
    self.build()
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1114, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1151, in _build
    build_save=build_save, build_restore=build_restore)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 795, in _build_internal
    restore_sequentially, reshape)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 406, in _AddRestoreOps
    restore_sequentially)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 862, in bulk_restore
    return io_ops.restore_v2(filename_tensor, names, slices, dtypes)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/ops/gen_io_ops.py", line 1466, in restore_v2
    shape_and_slices=shape_and_slices, dtypes=dtypes, name=name)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python
/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

NotFoundError (see above for traceback): Key Adam/beta_1 not found in checkpoint
         [[node save/RestoreV2 (defined at /home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/common/tf_util.py:301)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_INT64, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1556, in restore
    names_to_keys = object_graph_key_mapping(save_path)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1830, in object_graph_key_mapping
    checkpointable.OBJECT_GRAPH_PROTO_KEY)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 371, in get_tensor
    status)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/framework/errors_impl.py", line 528, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.NotFoundError: Key _CHECKPOINTABLE_OBJECT_GRAPH not found in checkpoint

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "action-branching-agents/agents/bdq/enjoy_continuous.py", line 84, in <module>
    main(args)
  File "action-branching-agents/agents/bdq/enjoy_continuous.py", line 53, in main
    act = deepq.load(args.model_dir)
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/deepq/procedure_continuous_tasks.py", line 77, in load
    return ActWrapper.load(path, num_cpu=num_cpu)
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/deepq/procedure_continuous_tasks.py", line 38, in load
    U.load_state(os.path.join(td, "model"))
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/common/tf_util.py", line 302, in load_state
    saver.restore(get_session(), fname)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1562, in restore
    err, "a Variable name or other graph key that is missing")
tensorflow.python.framework.errors_impl.NotFoundError: Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is missing from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key Adam/beta_1 not found in checkpoint
         [[node save/RestoreV2 (defined at /home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/common/tf_util.py:301)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_INT64, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]

Caused by op 'save/RestoreV2', defined at:
  File "action-branching-agents/agents/bdq/enjoy_continuous.py", line 84, in <module>
    main(args)
  File "action-branching-agents/agents/bdq/enjoy_continuous.py", line 53, in main
    act = deepq.load(args.model_dir)
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/deepq/procedure_continuous_tasks.py", line 77, in load
    return ActWrapper.load(path, num_cpu=num_cpu)
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/deepq/procedure_continuous_tasks.py", line 38, in load
    U.load_state(os.path.join(td, "model"))
  File "/home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/common/tf_util.py", line 301, in load_state
    saver = tf.train.Saver()
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1102, in __init__
    self.build()
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1114, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1151, in _build
    build_save=build_save, build_restore=build_restore)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 795, in _build_internal
    restore_sequentially, reshape)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 406, in _AddRestoreOps
    restore_sequentially)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 862, in bulk_restore
    return io_ops.restore_v2(filename_tensor, names, slices, dtypes)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/ops/gen_io_ops.py", line 1466, in restore_v2
    shape_and_slices=shape_and_slices, dtypes=dtypes, name=name)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/home/hbp/anaconda3/envs/rl_manip/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

NotFoundError (see above for traceback): Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is missing from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key Adam/beta_1 not found in checkpoint
         [[node save/RestoreV2 (defined at /home/hbp/rl_man/ActionBranchingDQN/action-branching-agents/agents/bdq/common/tf_util.py:301)  = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_INT64, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]_

**Comment by BarisYazici** [comment]
The problem is the TF variable scope when saving and loading the model. I use encoder for the observations of my environment. That's why, in the same TF session, there are encoder model variables and BDQ model variables. With the current model, it tries to save also my encoder's TF model variables.

I solved the problem by following the steps below.
1. I first updated the save and load function as in openai-baselines. 
https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/common/tf_util.py#L345

2. I update the save and load functionality to retrieve a specific variable scope. In our case, the scope is "deepq" so the line where they receive the variables changes to take only the "deepq" scope.
https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/common/tf_util.py#L348
`    variables = variables or tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope="deepq")`
